### PR TITLE
Clean up system_default_filename()

### PR DESCRIPTION
**Pull request by bstoeger**
In the old implementation there were two static C-style strings, filename
and path, which were initialized to NULL and filled on first call of
the function (i.e. singletons).

There is no sense in having two static variables indicating whether
this function was called previously. Moreover, there is no point
in remembering filename accross function calls, because it is not
used once path is set to a non-NULL value.

Therefore, make the filename variable non-static and calculate it only on
first invocation (as indicated by a NULL path). Moreover, free() the filename
variable after its use to fix a memory leak of the old code.

The windows code is slightly different in that the temporary filename is
not dynamically allocated.

Signed-off-by: Berthold Stoeger <bstoeger@mail.tuwien.ac.at>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

Perhaps I'm missing some subtleties, but this seems like an quite obvious clean up. It compiles and runs on Linux. Mac uses the same code, so should work likewise. Windows: let's wait for CI build....

### CODE REVIEW

**Review by neolit123 — CHANGES_REQUESTED** [review]
your commit message needs a SOB line.

this looks good to me, but i would leave another set of eyes to have a quick look.

@glance- @dirkhh @mturkia 
please, merge after the SOB is added.

**Review by neolit123 — APPROVED** [review]

File: core/linux.c [hunk 1]
```diff
@@ -83,18 +83,17 @@ const char *system_default_directory(void)
 
 const char *system_default_filename(void)
 {
-	static char *filename = NULL;
-	if (!filename) {
```
Comment (dirkhh) [inline core/linux.c:87]: one cannot comment on the commit message, which is a little hard to parse for me.
But obviously the filename can be NULL since it's a pointer.

File: core/linux.c [hunk 2]
```diff
@@ -83,18 +83,17 @@ const char *system_default_directory(void)
...
-		filename = calloc(strlen(user) + 5, 1);
+		char *filename = calloc(strlen(user) + 5, 1);
 		strcat(filename, user);
 		strcat(filename, ".xml");
-	}
-	static const char *path = NULL;
-	if (!path)
 		path = system_default_path_append(filename);
+		free(filename);
+	}
```
Comment (dirkhh) [inline core/linux.c:96]: And of course with this change here, the other PR that I closed a moment ago makes more sense as you changed how things are allocated. Would have been nice if the other PR had mentioned this one.

**Review by dirkhh — CHANGES_REQUESTED** [review]
I think I'd prefer to have the commit from the other PR that used copy_string when using system_default_filename() added to this PR so that the two changes are in the same place.
Also, please make sure that your commit message really describes the changes. At least to my reading the current one doesn't

File: core/linux.c (hunk 2 above)
Comment (bstoeger) [reply core/linux.c:96]: Sorry, but you're reading it wrong - these two PRs are perfectly independent. Note that it's *not* "filename" that is returned here, but "path". filename *always* leaked in this function.

If you want, I can combine both commits in a single PR, still your analysis is flawed.

**Comment by bstoeger** [comment]
I amended the commit message - please let me know if it is more clear.

Please also reconsider commit #831, which is in my opinion not dependent on this one here.

**Comment by dirkhh** [comment]
I haven't had coffee yet.
I'll have someone who is more awake deal with this and #831

**Comment by neolit123** [comment]
@mturkia @glance- @atdotde @janmulder 
requesting review from at least one more person.
